pkg/scheduler/api/hyper_node_info.go

```diff
@@ -0,0 +1,463 @@
+		hyperNodesListByTier: hyperNodesListByTier,
+		realNodesSet:         realNodesSet,
+		parentMap:            make(map[string]string),
+		ready:                ready,
+	}
+}
+
+// HyperNodeInfo stores information about a single HyperNode.
+type HyperNodeInfo struct {
+	Name      string
+	HyperNode *topologyv1alpha1.HyperNode
+	Tier      int
+
+	isDeleting bool
+}
+
+// NewHyperNodeInfo creates a new HyperNodeInfo instance.
+func NewHyperNodeInfo(hn *topologyv1alpha1.HyperNode, tier int) *HyperNodeInfo {
+	return &HyperNodeInfo{
+		Name:      hn.Name,
+		HyperNode: hn,
+		Tier:      tier,
+	}
+}
+
+// String returns a string representation of the HyperNodeInfo.
+func (hni *HyperNodeInfo) String() string {
+	return strings.Join([]string{hni.Name, strconv.Itoa(hni.Tier)}, "/")
+}
+
+// HyperNodes returns the map of all HyperNode information.
+func (hni *HyperNodesInfo) HyperNodes() map[string]*HyperNodeInfo {
+	return hni.hyperNodes
+}
+
+// HyperNodesListByTier returns a deep copy of the map that groups HyperNode names by their tier.
+// This ensures that the returned map is independent of the original, preventing unintended modifications.
+func (hni *HyperNodesInfo) HyperNodesListByTier() map[int]sets.Set[string] {
+	copiedHyperNodesListByTier := make(map[int]sets.Set[string], len(hni.hyperNodesListByTier))
+	for tier, row := range hni.hyperNodesListByTier {
+		copiedHyperNodesListByTier[tier] = row.Clone()
+	}
+
+	return copiedHyperNodesListByTier
+}
+
+// RealNodesSet returns a deep copy of the map that stores the sets of real nodes for each HyperNode.
+// This ensures that the returned map is independent of the original, preventing unintended modifications.
+func (hni *HyperNodesInfo) RealNodesSet() map[string]sets.Set[string] {
+	copiedRealNodesSet := make(map[string]sets.Set[string], len(hni.realNodesSet))
+	for name, value := range hni.realNodesSet {
+		copiedRealNodesSet[name] = value.Clone() // Clone the set to ensure a deep copy.
+	}
+
+	return copiedRealNodesSet
+}
+
+// ParentMap returns the map of parent-child relationships between HyperNodes.
+func (hni *HyperNodesInfo) ParentMap() map[string]string {
+	return hni.parentMap
+}
+
+// DeleteHyperNode deletes a HyperNode from the cache and update hyperNode tree.
+func (hni *HyperNodesInfo) DeleteHyperNode(name string) error {
+	hni.markHyperNodeIsDeleting(name)
+	if err := hni.updateAncestors(name); err != nil {
+		return err
+	}
+	hni.removeParent(name)
+
+	// We can safely delete hyperNode after updated ancestors.
+	hni.deleteHyperNode(name)
+	return nil
+}
+
+func (hni *HyperNodesInfo) UpdateHyperNode(hn *topologyv1alpha1.HyperNode) error {
+	hni.updateHyperNode(hn)
+	return hni.updateAncestors(hn.Name)
+}
+
+func (hni *HyperNodesInfo) BuildHyperNodeCache(hn *HyperNodeInfo, processed sets.Set[string], ancestors sets.Set[string], nodes []*corev1.Node) error {
+	// Check for cyclic dependency.
+	if ancestors.Has(hn.Name) {
+		hni.setReady(false)
```

Comment:
Already set `hni.setReady` in updateAncestors, remove this line?

Reply:
done